scripts/platform-init/platform-init.py

```diff
@@ -0,0 +1,202 @@
+            k8s_client.V1Secret,
+            v1.read_namespaced_secret(secret_name, namespace),
+        )
+        if not secret.data:
+            return None
+        secret_data = secret.data.get(secret_key)
+        if secret_data is not None:
+            secret_data = b64decode(secret_data).decode()
+        return secret_data
+
+    except k8s_client.ApiException:
+        return None
+
+
+def init_secret_service_secret(config: Config):
+    logging.info("Initializing secret service secret")
+    v1 = k8s_client.CoreV1Api()
+
+    private_key_secret = f"{config.renku_fullname}-secret-service-private-key"
+    private_key_entry_name = "privateKey"
+    existing_private_key = _get_k8s_secret(
+        config.k8s_namespace, private_key_secret, private_key_entry_name
+    )
+
+    public_key_secret = f"{config.renku_fullname}-secret-service-public-key"
+    public_key_entry_name = "publicKey"
+    existing_public_key = _get_k8s_secret(
+        config.k8s_namespace, public_key_secret, public_key_entry_name
+    )
+
+    if existing_private_key is None and config.secret_service_private_key is None:
+        # generate new secret
+        private_key = rsa.generate_private_key(public_exponent=65537, key_size=4096)
+        private_key_pem = private_key.private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.PKCS8,
+            encryption_algorithm=serialization.NoEncryption(),
+        )
+        v1.create_namespaced_secret(
+            config.k8s_namespace,
+            k8s_client.V1Secret(
+                api_version="v1",
+                data={private_key_entry_name: b64encode(private_key_pem).decode()},
+                kind="Secret",
+                metadata={
+                    "name": private_key_secret,
+                    "namespace": config.k8s_namespace,
+                },
+                type="Opaque",
+            ),
+        )
+    elif existing_private_key is None and config.secret_service_private_key is not None:
+        # create private key from config
+        private_key = serialization.load_pem_private_key(
+            config.secret_service_private_key.encode(), password=None
+        )
+        v1.create_namespaced_secret(
+            config.k8s_namespace,
+            k8s_client.V1Secret(
+                api_version="v1",
+                data={
+                    private_key_entry_name: b64encode(
+                        config.secret_service_private_key.encode()
+                    ).decode()
+                },
+                kind="Secret",
+                metadata={
+                    "name": private_key_secret,
+                    "namespace": config.k8s_namespace,
+                },
+                type="Opaque",
+            ),
+        )
+
+    elif (
+        existing_private_key != config.secret_service_private_key
+        and config.secret_service_private_key is not None
+    ):
+        # update private key
```

Comment:
Can we actually support updating the private key? What happens to all the secrets that are sitting in k8s secrets or in the database that are encrypted with the old key?

Same question for the public key.

Reply:
Also I think we should support two cases only:
1. Both private and public key are specified in the config and they are proper and valid
2. Neither public nor private key are specified in the config, in this case we generate the secret only if the secret does not already exist

Right now we are independently checking for the pivate and public keys and also generating them independently. I think this is too complicated and I dont think anyone would or should be using this option. I.e. provide a public key but let renku generate the private one.

Lastly there is no documentation on what algorithm/standard should be used for the keys. And I assume this matters or it does not? Like RSA, eliptic curve, etc...

Reply:
rotating keys is a nice to have that still needs to be handled.

I guess we could also just not require a public key at all and always generate it.

Reply:
> I guess we could also just not require a public key at all and always generate it.

I think that is fine

> rotating keys is a nice to have that still needs to be handled.

For this you would have to have some kind of watch in the secret service to restart it when the secret changes. I dont know how easy it is to implement this though. An easier way is to take the sha256 hash of the secret and make it an environment variable in the secret service. Then when the secret changes the secret service will automatically update. See https://helm.sh/docs/chart_template_guide/function_list/#sha256sum

Reply:
And taking the sha256 digest still does not address the issue of how you decrypt already encrypted secrets that may be still lying around. So perhaps it is better to keep this simple in the first implementation and improve this later during cooldown.